README.md
mobForest <img src="man/figures/200px-Rti-logo.png" align="right" />
========================================================

![travis-ci build status](https://travis-ci.org/RTIInternational/mobForest.svg?branch=master) ![CRAN statusbadge](https://www.r-pkg.org/badges/version/mobForest)

### mobForest R Package
`mobForest` implements random forest method for model based recursive partitioning. The `mob()` function, developed by Zeileis et al (2008), within party package, is modified to construct model-based decision trees based on random forests methodology. The main input function `mobforest.analysis()` takes all input parameters to construct trees, compute out-of-bag errors, predictions, and overall accuracy of forest. The algorithm performs parallel computation using `clusterApply()` function within the parallel package.

### Installation

```
install.packages("mobForest")
```

### Usage
To run the example, you will need the `mlbench` package. It contains a boston housing dataset for machine learning algorithms to run benchmark tests on. 

```
library(mlbench)
set.seed(1111)
# Random Forest analysis of model based recursive partitioning load data
data("BostonHousing", package = "mlbench")
BostonHousing <- BostonHousing[1:90, c("rad", "tax", "crim", "medv", "lstat")] 

# Recursive partitioning based on linear regression model medv ~ lstat with 3 trees.  1 core/processor used. 
rfout <- mobforest.analysis(as.formula(medv ~ lstat), c("rad", "tax", "crim"),
mobforest_controls = mobforest.control(ntree = 3, mtry = 2, replace = TRUE,
        alpha = 0.05, bonferroni = TRUE, minsplit = 25), data = BostonHousing,
        processors = 1, model = linearModel, seed = 1111)
rfout
```

NEWS.md
# Updating to R 3.4.3

Our latest version of mobForest was focused on revamping existing code to work with the newest version of R. We also made strides to consolidate all code to common formats. Although no substantive changes were made (no new functions or changes to algorithms), users should be aware of these smaller tweaks. 

* All mobForest functions now use the `period.sep` naming convention
* All mobForest specific parameters now use the `underscore_sep` naming convention
* Unit tests were written for many of the basic functions. Although several tests are not ran using `testthat`. These tests were causing `R CMD Check` to run indefinitely. Forums suggested that this happens when using the parallel package. All tests passed on the latest build when ran locally. 
* CRAN suggests not using `:::` when calling hidden functions of another package. To overcome this, we copied the necessary hidden packages from `party` into `utility.R`. 

# Updating to R 3.5.2

No changes.

R/allclass.R
#' Class \code{"mobforest.control"} of mobForest model
#'
#' Control parameters for random forest
#'
#'
#' @name mobforest.control-class
#' @docType class
#' @section Objects from the Class: Objects can be created by
#' \code{\link{mobforest.control}}.
#' @keywords classes
#' @examples
#'
#' # showClass("mobforest.control") The following code creates following forest
#' # controls: 400 trees to be constructed, sampling with replacement, a node
#' # contains at least 200 observations
#' mobforest_controls = mobforest.control(ntree = 400, mtry = 4,
#'     replace = TRUE, minsplit = 200)
#'
#' @export
setClass("mobforest.control",
         representation(ntree = "numeric", mtry = "numeric",
                        replace = "logical", fraction = "numeric",
                        mob_control = "list"),
         prototype = list(ntree = 300, mtry = 0,
                          replace = FALSE, fraction = 0.632))

#' Class \code{"varimp.output"} of mobforest model
#'
#' Variable importance
#'
#'
#' @name varimp.output-class
#' @docType class
#' @section Objects from the Class: Objects can be created by calls of the
#' form \code{\link{varimp.output}}. %% ~~ describe objects here ~~
#' @seealso \code{\linkS4class{varimp.output}}
#' @references Strobl, C., Malley, J. and Tutz, G. (2009) An introduction to
#' recursive partitioning: rationale, application, and characteristics of
#' classification and regression trees, bagging, and random forests,
#' \emph{Psychol Methods}, 14, 323-348.\cr
#' @export
setClass("varimp.output",
         representation(varimp_matrix = "matrix"),
         prototype = list(varimp_matrix = matrix(0, 0, 0)))


#' Class \code{"prediction.output"} of mobForest model
#'
#' The object of this class stores predictions and predictive accuracy
#' estimates.
#'
#'
#' @name prediction.output-class
#' @docType class
#' @section Objects from the Class: Objects can be created by calls of the
#' form \code{\link{prediction.output}}.
#'
#' @seealso \code{\link{prediction.output}}, \code{\link{predictive.acc}}
#' @keywords classes
#' @export
setClass("prediction.output",
         representation(pred_mat = "matrix",
                        R2_or_acc = "numeric",
                        mse = "numeric",
                        overall_r2_or_acc = "numeric",
                        pred_type = "character"),
         prototype = list(pred_mat = matrix(0, 0, 0),
                          R2_or_acc = numeric(),
                          mse = numeric(),
                          overall_r2_or_acc = numeric(),
                          pred_type = character()))


#' Class \code{"mobforest.output"} of mobforest model
#'
#' Random Forest output for model based recursive partitioning
#'
#'
#' @name mobforest.output-class
#' @aliases mobforest.output-class
#' get.pred.values, mobfores.output, logical, logical-method
#' mobforest.output-method
#' predictive.acc, mobforest.output, logical, logical-method
#' mobforest.output-method, mobforest.output-method
#' mobforest.output-method
#' @docType class
#' @section Objects from the Class: Objects can be created by
#' \code{\link{mobforest.output}}.
#' @seealso \code{\linkS4class{prediction.output}},
#' \code{\linkS4class{varimp.output}}
#' @keywords classes
#' @rdname mobforest.output-class
#' @examples
#'
#' \dontrun{
#' library(mlbench)
#' set.seed(1111)
#' # Random Forest analysis of model based recursive partitioning load data
#' data("BostonHousing", package = "mlbench")
#' BostonHousing <- BostonHousing[1:90, c("rad", "tax", "crim", "medv", "lstat")]
#' 
#' # Recursive partitioning based on linear regression model medv ~ lstat with 3
#' # trees.  1 core/processor used. 
#' rfout <- mobforest.analysis(as.formula(medv ~ lstat), c("rad", "tax", "crim"),
#'     mobforest_controls = mobforest.control(ntree = 3, mtry = 2, replace = TRUE,
#'         alpha = 0.05, bonferroni = TRUE, minsplit = 25), data = BostonHousing,
#'     processors = 1, model = linearModel, seed = 1111)
#' }
#'
#' @exportClass mobforest.output
setClass("mobforest.output",
         representation(oob_predictions = "prediction.output",
                        general_predictions = "prediction.output",
                        new_data_predictions = "prediction.output",
                        varimp_object = "varimp.output",
                        model_used = "character",
                        family = "character",
                        train_response = "data.frame",
                        new_response = "data.frame"
                        ))
